src/Ads/AssetSuggestionsService.php

```diff
@@ -227,19 +300,76 @@ protected function get_shop_products( array $args = [] ): array {
+	 * Get gallery images ids.
 	 *
-	 * @param array $ids A list of attachments ids.
+	 * @param int $post_id Post ID that contains the gallery.
 	 *
-	 * @return array A list of attachments urls.
+	 * @return array List of gallery images ids.
+	 */
+	protected function get_gallery_images_ids( int $post_id ): array {
+		$gallery = get_post_gallery( $post_id, false );
+
+		if ( ! $gallery || ! isset( $gallery['ids'] ) ) {
+			return [];
+		}
+
+		return explode( ',', $gallery['ids'] );
+	}
+
+	/**
+	 * Get unique attachments ids converted to int values.
+	 *
+	 * @param array $ids Attachments ids.
+	 * @param int   $maximum_images Maximum number of images to return.
+	 *
+	 * @return array List of unique attachments ids converted to int values.
 	 */
-	protected function get_url_attachments_by_ids( array $ids ): array {
+	protected function prepare_image_ids( array $ids, int $maximum_images = self::DEFAULT_MAXIMUM_MARKETING_IMAGES ): array {
 		$ids = array_unique( ArrayUtil::remove_empty_values( $ids ) );
+		$ids = array_map( 'intval', $ids );
+		return array_slice( $ids, 0, $maximum_images );
+	}
+
+	/**
+	 * Get URL for each attachment using an array of attachment ids and a list of subsizes.
+	 *
+	 * @param array $ids Attachments ids.
+	 * @param array $size_keys Image subsize keys.
+	 * @param int   $maximum_images Maximum number of images to return.
+	 *
+	 * @return array A list of attachments urls.
+	 */
+	protected function get_url_attachments_by_ids( array $ids, array $size_keys = [ self::SQUARE_MARKETING_IMAGE_KEY, self::MARKETING_IMAGE_KEY ], $maximum_images = self::DEFAULT_MAXIMUM_MARKETING_IMAGES ): array {
+		$ids = $this->prepare_image_ids( $ids, $maximum_images );
 
 		$marketing_images = [];
+
 		foreach ( $ids as $id ) {
-			$marketing_images[] = wp_get_attachment_image_url( $id );
+
+			$metadata = wp_get_attachment_metadata( $id );
+
+			foreach ( $size_keys as $size_key ) {
+
+				$minimum_size     = new DimensionUtility( ...$this->image_requirements[ $size_key ]['minimum'] );
+				$recommended_size = new DimensionUtility( ...$this->image_requirements[ $size_key ]['recommended'] );
+				$image_size       = new DimensionUtility( $metadata['width'], $metadata['height'] );
+				$suggested_size   = $this->image_utility->recommend_size( $image_size, $recommended_size, $minimum_size );
+
+				// If the original size matches the suggested size with a precision of +-1px.
+				if ( $suggested_size && $suggested_size->equals( $image_size ) ) {
+					$marketing_images[ $size_key ][] = wp_get_attachment_url( $id );
+				} elseif ( isset( $metadata['sizes'][ $size_key ] ) ) {
+					// use the sub size.
+					$marketing_images[ $size_key ][] = wp_get_attachment_image_url( $id, $size_key );
+				} elseif ( $suggested_size && $this->image_utility->maybe_add_subsize_image( $id, $size_key, $suggested_size ) ) {
+					// use the resized image.
+					$marketing_images[ $size_key ][] = wp_get_attachment_image_url( $id, $size_key );
```

Comment:
💅  Optionally this 2 conditionals could be merged into one

```php
// If the original size matches the suggested size with a precision of +-1px.
if ( $suggested_size && $suggested_size->equals( $image_size ) ) {
	$marketing_images[ $size_key ][] = wp_get_attachment_url( $id );
} elseif ( isset( $metadata['sizes'][ $size_key ] ) || ( $suggested_size && $this->image_utility->maybe_add_subsize_image( $id, $size_key, $suggested_size ) ) ) {
	// use the previously created subsize  or create a new subsize.
	$marketing_images[ $size_key ][] = wp_get_attachment_image_url( $id, $size_key );
}
```

Reply:
This is just an alternative. Separate in different conditions groups also ok.

Reply:
I will keep it as it is, as I think it is a bit easier to read it.